### PR TITLE
Track 1: Add FP8 MLP down projection (-1.04s)

### PR DESCRIPTION
**Pull request by Mister-dev-oss**
## Summary

This PR builds on the current Track 1 `master` record and adds one systems-side
change to `train_gpt.py` and `triton_kernels.py`:

**FP8 MLP down projection.** The MLP down-projection forward path now runs in
FP8 via `torch._scaled_mm`. The existing fused up-projection/ReLU-squared Triton
kernel also emits FP8 post-activation values and partial activation amax values.
The MLP down-projection weights are quantized and transposed to FP8 with delayed
scales. The backward path still uses the original BF16 activations and BF16
weights.

The standard run command is unchanged:

```bash
python data/cached_fineweb10B.py 10
torchrun --standalone --nproc_per_node=8 train_gpt.py
```

## What Changed

- Add FP8 post-activation emission to the fused up-projection/ReLU-squared
  Triton kernel.
- Add delayed activation-scale reduction for the FP8 MLP down-projection input.
- Add FP8 quantize+transpose for the MLP down-projection weights.
- Route the forward down projection through `torch._scaled_mm`.
- Keep the backward path in BF16, using the same saved BF16 tensors as before.

## Main Result

Both baseline and this PR were run on the same 8xH100 setup with the same
environment:

- PyTorch `2.10.0+cu128`
- Triton `3.6.0`
- CUDA `12.8`
- 13 runs per variant

| Run | Mean val loss | Loss std | p(mean < 3.28) | Mean train time | Time std |
| --- | ---: | ---: | ---: | ---: | ---: |
| Current `master` baseline | 3.2782461538 | 0.0020164389 | 0.0042976481 | 80.728615s | 0.065165s |
| This PR | 3.2783692308 | 0.0014156180 | 0.0006690604 | 79.684923s | 0.182767s |

Logs and per-variant statistics are included in:

- `records/track_1_short/2026-07-17_FP8DownProjection/baseline/`
- `records/track_1_short/2026-07-17_FP8DownProjection/this_pr/`

## Delta

Relative to the same-hardware baseline:

- Mean train time: `79.684923s - 80.728615s = -1.043692s`
- Relative train time: `-1.292841%`
- Speedup: `1.01309774x`
- Mean val loss: `3.2783692308 - 3.2782461538 = +0.0001230769`

The loss distributions are statistically indistinguishable in the baseline-vs-PR
comparison:

- Welch p-value for candidate loss vs baseline loss: `0.8587465`

The timing separation is stable across the 13 runs:

- Welch t-statistic for timing: `19.394`
- Welch p-value for timing: `4.9e-12`
- Slowest candidate run: `80.231s`
- Fastest baseline run: `80.610s`

## Validity

**Target p-value passes.** This PR reaches `p = 0.0006690604` for
`mean val loss < 3.28`, below the Track 1 `p < 0.01` requirement.

**Same validation path.** Validation still uses the standard model path and
validation data. The FP8 down-projection change only affects the training
forward MLP down projection.

**Same data and batching.** The data pipeline is unchanged. The run uses the
standard FineWeb cached data preparation and the default Track 1 batch/sequence
configuration from `train_gpt.py`.

**Same command and compile settings.** The standard `torchrun` command is
unchanged, and this PR does not add extra `torch._inductor.config` or
`torch.compile` flags.

**Same hardware comparison.** Baseline and candidate were both run on the same
8xH100 setup with separate compiler caches and 13 fresh runs each.

**Backward path unchanged in precision.** Backward still consumes BF16
activations and BF16 down-projection weights. The change is a forward-path
systems optimization.

## Run Data

Baseline losses:

```python
[3.2817, 3.2796, 3.2778, 3.2802, 3.2775, 3.2767, 3.2758, 3.2821, 3.2780, 3.2768, 3.2769, 3.2775, 3.2766]
```

Baseline times:

```python
[80.610, 80.710, 80.678, 80.720, 80.747, 80.693, 80.659, 80.694, 80.763, 80.795, 80.818, 80.840, 80.745]
```

This PR losses:

```python
[3.2791, 3.2773, 3.2783, 3.2775, 3.2781, 3.2775, 3.2793, 3.2772, 3.2769, 3.2775, 3.2806, 3.2779, 3.2816]
```

This PR times:

```python
[79.536, 79.740, 79.814, 79.658, 79.593, 79.629, 80.231, 79.653, 79.641, 79.532, 79.539, 79.682, 79.656]
```

Calculation snippet:

```python
import scipy.stats
import torch

for name, losses, times in [
    ("baseline", baseline_losses, baseline_times),
    ("this_record", this_record_losses, this_record_times),
]:
    print(name)
    print("p=%.10f" % scipy.stats.ttest_1samp(losses, 3.28, alternative="less").pvalue)
    print("losses:", torch.std_mean(torch.tensor(losses)))
    print("time:", torch.std_mean(torch.tensor(times)))
```

## Acknowledgements

This work was made possible thanks to the support provided by Paradigma ([link](https://paradigma.inc/)) and Flywheel
([link](https://flywheel.paradigma.inc/)), Paradigma's infrastructure for research.

Team: [@Mister-dev-oss](https://github.com/Mister-dev-oss),[@CerovazS](https://github.com/CerovazS),[@MarioPaerle](https://github.com/MarioPaerle),[@GabrieleCirillo](https://github.com/GabrieleCirillo),[@crisostomi](https://github.com/crisostomi).

### CODE REVIEW

**Comment by msisovic** [comment]
Great job! I also tried adding this, had pretty similar setup, with delayed scaling and forward-only, when adding the fp8 up-proj, but didn't manage to pull it off. Excited to take a look at how you did it.

**Comment by ClassicLarry** [comment]
Awesome. Very advanced FP8 implementation. The delayed scaling metric is clever.

Merging at 73.8s (-0.8s), based on timing on main.
